### DISCUSSION TITLE
Support Vue with the same DX than React

### DISCUSSION THREAD

**Discussion by JSteunou** [lingui/js-lingui#1730] [General]
I'm starting to work with Vue, coming from the React world, and I really want to keep using lingui because I think it is the best frontend solution for handling i18n issues.

Lingui + Vue + Vite works with the basics, as a JS project, so atm you do not have some features like : 

* id generation
* Components
* handling html tags in your sentences with placeholders / children components.
* ...

In fact every goodies that make Lingui having a very good DX.

I want to make this happen, but I need the help of Vue & Lingui masters here.

* For the JS/TS/setup part I though about using macros, which should be split in order to not import the React universe.
* For the template part I though about relying on the vue-extractor + AST reading for the extraction part, and a set of Vue components using render function for the rendering part.

Who's on board?

**Comment by SpiderDan98** [comment]
Here are hopefully some good links:

- https://github.com/lingui/js-lingui/discussions/1017
- https://github.com/lingui/js-lingui/issues/1361

**Comment by timofei-iatsenko** [comment]
I'm not very familiar with Vue, so i could not help on the best solution in that ecosystem.

You already can use just `t` and `msg` macros, and it will give auto-id generation, ICU expanding and so on. 

To replicate JSX macro in Vue templates you need to somehow hook into the Vue compiler, and as far as i understand it's not something very popular and doable.  So the easiest way would be just pass strings created using macro from setup script to template.

**Reply by JSteunou** [reply]
And that's I'll gonna do most of the time, but there is this tiny but recurrent use case: html tag inside sentences. Just for that one, I miss the React babel macro a loooooot!

**Reply by JSteunou** [reply]
Another simple thing is the documentation. As of today I cannot get lingui fully work with my stack: there is one tutorial for vite with react, another for the vue extractor, but not a simple way for vite + vue. As of today, template aside, I still do not have a fully working flow with lingui + vite + vue.

**Reply by SpiderDan98** [reply]
So that's how lingui runs with vite and vue. Everything should work, except for the macros. 
https://codesandbox.io/p/sandbox/serene-forest-5gr5gr

**Reply by SpiderDan98** [reply]
I got the macros to work inside the script tag. Vue doesn't transform the script code with babel. The code is just parsed with babel. So I added a babel `transform` to the vue compiler. I think it's dirty, but it works.
Note: It doesn't work inside the template.

```ts
    vue({
      compiler: {
        ...compiler,
        parse: (source, options) => {
          const scripts = source.match(
            /<script([\s\S]*?)>([\s\S]*?)<\/script>/gms
          );

          for (const script of scripts) {
            const startIndex = script.indexOf(">") + 1;
            const endIndex = script.lastIndexOf("<");

            const startTag = script.substring(0, startIndex);
            const endTag = script.substring(endIndex, script.length);

            const scriptContent = script.substring(startIndex, endIndex);

            const { code } = transform(scriptContent, {
              plugins: ["macros"],
            });

            source = source.replace(script, `${startTag}${code}${endTag}`);
          }

          return compiler.parse(source, options);
        },
      },
    }),
```

**Reply by timofei-iatsenko** [reply]
Here is better solution without string manipulation to support macro in script tags. 

https://github.com/lingui/js-lingui/issues/1567#issuecomment-1503761631

**Reply by JSteunou** [reply]
I got macro running with vite + vuejs starting from `npm create vue@3` with your method @thekip,  but it only works with build + preview not dev mode (hmr yelling red). Any clue?

**Reply by timofei-iatsenko** [reply]
Have no idea, need to investigate. Might be related to the phase where babel macro transformation is applied (before / after HMR injection).

**Reply by JSteunou** [reply]
I have this config, I would not pull all the react / jsx deps that's why it is different than the one you shared. But maybe I broke something.

```js
const sourceRegex = /\.(:?[j|t]sx?|vue)$/ // <-- accept vue files 

function macrosPlugin(): Plugin {
  return {
    name: "babel-macros",
    async transform(source: string, filename: string) {
      if (filename.includes("node_modules")) {
        return undefined
      }
      
      if (!sourceRegex.test(filename)) {
        return undefined
      }

      const result = await babel.transformAsync(source, {
        filename,
        plugins: [
          // add your babel plugins here
          require.resolve("babel-plugin-macros"),
        ],
        babelrc: false,
        configFile: false,
        sourceMaps: true,
      })
      return result as TransformResult | null
    },
  } as const
}

```

**Reply by JSteunou** [reply]
I find out that the issue happens only when using `i18n.t` inside template section, but all goes well when using `i18n._`

This is the error I have.

```
12:08:02 [vite] hmr update /src/components/HelloWorld.vue
12:08:02 [vite] Internal server error: /home/jerome/Github/poclingui_vue/reproduction/src/components/HelloWorld.vue: @lingui/macro: Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros. 
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues
 
 Error: Property value of ObjectProperty expected node to be of a type ["Expression","PatternLike"] but instead got "ReturnStatement"
  19 |       return bar;
  20 |     }, get t() {
> 21 |       return t;
     |       ^^^^^^^^^
  22 |     }, get i18n() {
  23 |       return i18n;
  24 |     } }; Learn more: https://www.npmjs.com/package/@lingui/macro
  Plugin: babel-macros
  File: /home/jerome/Github/poclingui_vue/reproduction/src/components/HelloWorld.vue
      at File.buildCodeFrameError (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/transformation/file/file.js:205:12)
      at NodePath.buildCodeFrameError (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/index.js:105:21)
      at reportUnsupportedSyntax (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@lingui/macro/dist/index.cjs:843:14)
      at /home/jerome/Github/poclingui_vue/reproduction/node_modules/@lingui/macro/dist/index.cjs:823:7
      at Array.forEach (<anonymous>)
      at macro (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@lingui/macro/dist/index.cjs:813:49)
      at macroWrapper (/home/jerome/Github/poclingui_vue/reproduction/node_modules/babel-plugin-macros/dist/index.js:57:12)
      at applyMacros (/home/jerome/Github/poclingui_vue/reproduction/node_modules/babel-plugin-macros/dist/index.js:226:14)
      at ImportDeclaration (/home/jerome/Github/poclingui_vue/reproduction/node_modules/babel-plugin-macros/dist/index.js:110:28)
      at NodePath._call (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:46:20)
      at NodePath.call (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:36:17)
      at NodePath.visit (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:82:31)
      at TraversalContext.visitQueue (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:86:16)
      at TraversalContext.visitMultiple (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:61:17)
      at TraversalContext.visit (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:107:19)
      at traverseNode (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/traverse-node.js:18:17)
      at traverse (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/index.js:49:34)
      at NodePath.traverse (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/index.js:108:24)
      at PluginPass.Program (/home/jerome/Github/poclingui_vue/reproduction/node_modules/babel-plugin-macros/dist/index.js:91:18)
      at newFn (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/visitors.js:159:14)
      at NodePath._call (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:46:20)
      at NodePath.call (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:36:17)
      at NodePath.visit (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/path/context.js:82:31)
      at TraversalContext.visitQueue (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:86:16)
      at TraversalContext.visitSingle (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:65:19)
      at TraversalContext.visit (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/context.js:109:19)
      at traverseNode (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/traverse-node.js:18:17)
      at traverse (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/traverse/lib/index.js:49:34)
      at transformFile (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/transformation/index.js:81:29)
      at transformFile.next (<anonymous>)
      at run (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/transformation/index.js:24:12)
      at run.next (<anonymous>)
      at transform (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/transform.js:22:41)
      at transform.next (<anonymous>)
      at step (/home/jerome/Github/poclingui_vue/reproduction/node_modules/gensync/index.js:261:32)
      at evaluateAsync (/home/jerome/Github/poclingui_vue/reproduction/node_modules/gensync/index.js:291:5)
      at /home/jerome/Github/poclingui_vue/reproduction/node_modules/gensync/index.js:93:9
      at new Promise (<anonymous>)
      at async (/home/jerome/Github/poclingui_vue/reproduction/node_modules/gensync/index.js:92:14)
      at stopHiding - secret - don't use this - v1 (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/errors/rewrite-stack-trace.js:47:12)
      at Object.transformAsync (/home/jerome/Github/poclingui_vue/reproduction/node_modules/@babel/core/lib/transform.js:46:77)
      at TransformContext.transform (/home/jerome/Github/poclingui_vue/reproduction/vite.config.ts:64:34)
      at Object.transform (file:///home/jerome/Github/poclingui_vue/reproduction/node_modules/vite/dist/node/chunks/dep-abb4f102.js:44206:62)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async loadAndTransform (file:///home/jerome/Github/poclingui_vue/reproduction/node_modules/vite/dist/node/chunks/dep-abb4f102.js:54851:29)
      at async viteTransformMiddleware (file:///home/jerome/Github/poclingui_vue/reproduction/node_modules/vite/dist/node/chunks/dep-abb4f102.js:64232:32)
```

It seems using macro + core along is not ideal. Logic for macro try to replace the `i18n.t` call?

**Reply by timofei-iatsenko** [reply]
Show me the input code. This not a related to macro, it's related to how Vue compiler transforms the code.  It tries to push `t` variable from macro (which is virtual) to the template function, you can clearly see it here: 

```
 19 |       return bar;
  20 |     }, get t() {
> 21 |       return t;
     |       ^^^^^^^^^
  22 |     }, get i18n() {
  23 |       return i18n;
```

And then babel-macro plugin attempt to parse this and fail because there are no handlers in lingui-macro for such cases.

**Reply by JSteunou** [reply]
You are right. After digging I had the same feeling. Here is what looks like the code after vue compiler / before macro in dev mode.

* :negative_squared_cross_mark: extract (known, because of template)
* :negative_squared_cross_mark: dev mode HMR
* :part_alternation_mark: build pass but component render is broken, empty view

```js
    _createElementVNode(
      "h2",
      null,
      _toDisplayString($setup.t`inside a t macro`),
      1
      /* TEXT */
    ),
    _hoisted_3
  ]);
```

Vue links every template *global* variable to `$setup` var.

This one also fails (in HMR) 

* :heavy_check_mark: extract is ok
* :negative_squared_cross_mark: dev mode HMR (same error with macro)
* :heavy_check_mark: build & preview works



```js
    _createElementVNode(
      "h2",
      null,
      _toDisplayString($setup.i18n.t("inside a t core i18n")),
      1
      /* TEXT */
    ),
    _hoisted_3
  ]);
```

This one goes well

* :heavy_check_mark: extract is ok
* :heavy_check_mark:  dev mode HMR works
* :heavy_check_mark: build & preview works

```js
    _createElementVNode(
      "h2",
      null,
      _toDisplayString($setup.i18n._("inside a _ core i18n")),
      1
      /* TEXT */
    ),
    _hoisted_3
  ]);
```


Update: list all cases

**Reply by JSteunou** [reply]
During a build, vue wraps those setup var with `_unref` which make the macro fails

in

```js
        _createElementVNode("h2", null, _toDisplayString(_unref(t)`inside a t macro`), 1),
```

out

```js
_createElementVNode("h2", null, _toDisplayString(i18n._(
      /*i18n*/
      {
        id: "J/hVSQ",
        values: {
          0: _unref(t)
        }
      })`inside a t macro`), 1),
```

**Reply by timofei-iatsenko** [reply]
And because macro used in template is referenced not directly from `import`, but passed thru  `$setup` var it's hardly possibly to find it using static analysis and expand the macro. The vue compiler produces the bindings metadata during compilation (i playd with a compiler a bit while helped to implement vue extractor), but i didn't find a way how to hook into compiler in vite-vue and tight all parts together.

**Reply by JSteunou** [reply]
Yup‚ so far I see there is no choice to play with vue compiler. Doable in vue extractor‚ to go from a Trans component to an i18n._ call with id generation and maybe as vite plugin before babel macro

**Comment by JSteunou** [comment]
Following the discussion above, we now are sure, as of today, using vue@3 + vite + lingui & extractor & vite plugin, it is possible to:

* use any lingui feature inside SFC setup part (macro or core)
* use lingui core features inside SFC template part (but call to i18n.t break the HMR in dev mode)

The goal would be to improve the DX inside the template part and for that I think it is doable to plug into the lingui extractor for vue between the vue parser and the vue compiler in order to update the template AST.

Going from

```vue
<template>
  <Trans>Hi, I am a <b>statement</b> withing my {{ name }} template</Trans>
</template>
```

to

```vue
<template>
  {{ i18n._({id: "XXXX", message: "Hi, I am a <0>statement</0> withing my {name} template", comments: "", context: ""}) }}
</template>
```

I still need to found in vue how the descriptor can be updated if and when the AST is changed

Next step would be to reuse part of that logic to transform the template via vite to produce a real Trans component code with slots for children.

**Reply by JSteunou** [reply]
Using `nodeTransforms` from the `compilerOptions` I was able to write a quick PoC that use the AST & update it to transform a `<Trans>` into a `{{ i18n._ }}` and make the extractor fully extract sentences with markup & interpolation in it.

I'm looking for the part in lingui that generate the id @thekip do you know if and where it is exported by any chance?

**Reply by andrii-bodnar** [reply]
@JSteunou please check the [`@lingui/message-utils`](https://github.com/lingui/js-lingui/tree/main/packages/message-utils) package

**Reply by JSteunou** [reply]
I again, I'm back at it after several months :)

I now can extract and use lingui inside vue template section using fake Trans component processed during vue pre-compilation, but I cannot get the exact same experience than native lingui.

The issue is around id generation: my code transform `<Trans>some message</Trans>` into `i18n._(...)` code with message descriptor, but if I provide an id it is considered as explicit id, and if I don't, the message is the id. Either case I cannot get lingui handle the message id *under the hood* as it does with macro.

Any idea to achieve this?

**Reply by JSteunou** [reply]
I might generated the id the wrong way, now I can get it right.

@thekip or @andrii-bodnar can you confirm than when passing an id to i18n._ call, during extration, if lingui generate the same id in core it is considered as *the* internal id and not a custom message id?

**Reply by timofei-iatsenko** [reply]
if hash generated from message using [@lingui/message-utils](https://github.com/lingui/js-lingui/tree/main/packages/message-utils) is the same as provided ID -  it's considered "internal"

**Comment by JSteunou** [comment]
After several months away from this project I'm back and I have a thing working for vue template. It is not complete, it is a start with the `<Trans>` component and support for interpolation and html tag in it. Other components like Plural should be easy to add now.

How I did it: I wrote an extractor based on the actual lingui vue extractor but with a nodeTransforms for the SFC compilerTemplate part and another transformer for the same SFC template compilerOptions for the vite vue plugin (but can easily be used anywhere I think). Both transformers have a lot of code in common obviously.

How it works ?
* during the extraction it transforms all the `<Trans>...</Trans>` it can find (reading the AST) into `i18n._()` call
* during the build it transforms all the `<Trans>...</Trans>` it can find (still reading the AST) into `<Trans id="" message="" values=""></Trans>` with template slots as children when needed instead of the original message. Kudos to [Ugzuzg ideas](https://github.com/Ugzuzg/vue-lingui/blob/master/examples/hello/src/components/HelloWorld.vue)
* during the runtime the Trans component uses the id, message, values from the props and then format the result with the slots in a render function using a formater inspired by [the one from lingui react](https://github.com/lingui/js-lingui/blob/main/packages/react/src/format.ts)

What still need to be done?
* publishing it obviously: @thekip or any lingui maintainer I need your guidance here :sweat_smile: 
* adding Plural, Select, formater, ... support
* use transform to update the import so we can have 2 Trans component, one acting like a macro, the other the real one (as of today I cheat a little, only one component, all props are optional)
* documentation
* tests

**Reply by andrii-bodnar** [reply]
@JSteunou this sounds awesome!

Could you please push your code to some repository to showcase the extractor and transformer?

We would love to see it and discuss further publishing actions.

**Reply by JSteunou** [reply]
Sure thing. I'm currently integrating it in our monorepo first, kind of alpha testing it, then I will invite you both on a separate copy.

**Reply by JSteunou** [reply]
Done. I will keep it private until it reaches an acceptable state, then I will switch it to public and / or you could add it into lingui project.

To be honest I will be using it at my job so every issue I'll find I'll fix it, but I'm not sure to be able to maintain it on my own fully. So I think it would be better to integrate it into lingui project.

**Reply by JSteunou** [reply]
Some news from the battlefield:

* I added up a helper `vt` that does the same thing (but not yet all the things) that `t` macro does. So translation in template tag attributes works
* I found a better way to handle inner tags (and inner of inner of inner ...) that gives supa powa to `<Trans>` and allow unlimited inner tags and components depth
* I added some usage & explanation doc in the README

The big part that still need to be done is unit testing (yes I'm lazy I dev all the thing through example & demo case)

**Reply by andrii-bodnar** [reply]
Hi @JSteunou,

Thanks a lot for your work! I just checked out the repository and was amazed at the progress you've made.

I think it's a great addition to the Lingui project and I'm happy to help you with the release process.

I wonder what the next steps might be to make it happen. Probably we can include it as a separate package in the current repository, and also move the demo project to the `examples` folder.

I think we should create a separate article in the documentation that covers all aspects of working with Vue. Details about the technical implementation can be left in the README of the package.

From my perspective, the plan might look like this:

- finish the implementation of the `<Trans>` component
- finish the "support messages as prop during extraction" (maybe not completely, but at least the most critical parts)
- add tests
- prettify the package and the demo project, match the style of the other packages
- add documentation
- release

Also, I noticed some code duplication from the `@lingui/extractor-vue` package. What would be the best way to handle this? It would be great to avoid having two similar extractors.

What do you think about this?

**Reply by JSteunou** [reply]
Thank you :)

Totally agree with your plan.

* The <Trans> component is 80% complete, the other features can be added one by one later to make it 100% similar to the React one
* regarding the prop you might talk about the  `vt` helper I added. It works very well today with some differences with the `t` helper from lingui macro / react. No i18n instance can by given in parameter, no context, and all variables are extracted / replaced by index (0, 1, ...), because I choose the shortest path and make `vt` run as is in runtime, without transformation, thanks to `generateMessageId` available for browsers. Also no complexes forms, not inside `<Trans>` inner tag. Very simple support but I think it covers 90% of use cases.
* tests are currently in progress and will be added by the end of the week

About `@lingui/extractor-vue` I think it should be deprecated as the new one superseed it. 

Then about the integration I'm wondering if it should be as it is now or if module naming and folders should be rethink. Do you think one package is ok, or should it be split in multiple packages (extractor, vue, ...)

Maybe we should also start to think about and anticipate a split / renaming of `@lingui/macro` to have react-macro, vue-macro, etc.
